sdk/appconfiguration/azure-spring-cloud-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/CalculatedBackoffTime.java

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.spring.cloud.config;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.concurrent.TimeUnit;
+
+import com.azure.spring.cloud.config.properties.AppConfigurationProviderProperties;
+
+/**
+ * Calculates the amount of time to the next refresh, if a refresh fails.
+ */
+public final class CalculatedBackoffTime {
+
+    /**
+     * Calculates the amount of time to the next refresh, if a refresh fails. Takes current Refresh date into account
+     * for watch keys.
+     * @param state state being checked
+     * @param properties App Configuration Provider Properties
+     * @return new Refresh Date
+     */
+    public static Instant calculateBefore(State state, AppConfigurationProviderProperties properties) {
+        // The refresh interval is only updated if it is expired.
+        if (Instant.now().isAfter(state.getNextRefreshCheck())) {
+            Integer attempt = state.getRefreshAttempt();
+            state.addRefreshAttempt();
+
+            return calculate(attempt, state.getRefreshInterval(), properties);
+        }
+
+        return state.getNextRefreshCheck();
+    }
+
+    /**
+     * Calculates the amount of time to the next refresh, if a refresh fails. Takes current Refresh date into account
+     * for watch keys. Used for checking client refresh-interval only.
+     * @param nextRefreshCheck next refresh for the whole client
+     * @param attempt refresh attempt for the client
+     * @param interval the Refresh Interval
+     * @param properties App Configuration Provider Properties
+     * @return new Refresh Date
+     */
+    public static Instant calculateBefore(Instant nextRefreshCheck, Integer attempt, Duration interval,
```

Comment:
If this is exclusively used for calculating backoff time for client refresh, we can rename both `calculateBefore` methods to clarify this behavior. How about `getNextClientRefresh` and `getNextWatchKeyRefresh`?
Also suggest renaming the class to `BackoffTimeCalculator`.

I think usage will be easier to understand: 
`Instant nextRefreshTime = BackoffTimeCalculator.getNextClientRefresh(...)`
`Instant nextRefreshTime = BackoffTimeCalculator.getNextWatchKeyRefresh(...)`

Reply:
**Update after full review**: 
- We should have just one method for both watch keys and client refresh. 
- The method should return backoff time interval, not `Instant.now().plusSeconds(backoffInterval)`. 
- The method can be called `getNextBackoff()`. 
- The class can be called `BackoffTimeCalculator`.

Finally, the usage should look like this:
```
backoffInterval = BackoffTimeCalculator.getNextBackoff(...);
nextRefreshTime = Instant.now().plusSeconds(backoffInterval);
```

Reply:
#Resolved